src/preview/src/components/AdPreview.js

```diff
@@ -96,6 +138,7 @@ export const AdPreview = (props) => {
               <IconButton
                 onClick={(e) => {
                   activeConfigTab === "html,iframe" ? (adPreviewCard.current.src = ad.output.html.url) : setActiveConfigTab("html,iframe");
+                  setReload(v => !v);
```

Comment:
Can we really toggle true/false? What if you have more than two tabs? (gif, mp4, jpg etc)
@mm-vasyl

Reply:
this is just to trigger React to update the iframe effects on the reload button click. It always goes to HTML so just 1 tab to consider.

Reply:
And re odd-looking code:
I was testing all these in the `MINI template`, where things are a bit different, but the main logic is the same:

Since React renders things in queues (so not instantly) we can't rely on `onload` event. We must have some `trigger` to tell us when both sides are ready. So message communication and Promises come to help.

That's why we need this `const previewReady` Promise trigger. And we must define it **before** the async `init`.

So the code for the generator template would look more like this:
main.js
```
const previewReady = new Promise((resolve) => {
  window.addEventListener('previewReady', resolve);
});

banner.init().then(() => {
  
  banner.setAnimation(new Animation(document.querySelector('.banner'), config));
 
  banner.start();
  
  previewReady.then(() =>
    window.dispatchEvent(new CustomEvent('getMainTimeline', { detail: banner.animation.getTimeline() })),
  );
});
```

Or just update the `display-temple` so you don't need to add anything
This example is when you don't use `FrameAnimation` class or `frameN` functions

And the note that 'it works even without correct order':
Yes, but sometimes it doesn't, sometimes we're 1-2ms ahead of preview (React rendering) with onload and then it doesn't. So we should validate and align both with the Promise before init.

Hope it is clear

Reply:
https://github.com/mediamonks/display-dev-server/assets/45459258/0bb6a77b-9de0-4159-8dc8-c6d35da2f018

Same issue when trying via the new display-temple :) (don't mind the other errors, that is a chrome extension bug)

Reply:
oops, error on a preview side, forgot to check if we're in iframe or not. Fixed now